From 2824992d3ad7c35e063427736ffe11121176f444 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 27 Jun 2022 21:29:51 +0300
Subject: [PATCH] [SW-0] gaudi3: add minimum size dma check

Currently, there is no minimum dma check in dma copy function which
allows copy of dma with 0 size. This patch add required check
and returns error if minimum dma size is not passed to API.

Change-Id: I035c9302b56552ebfa7d7635c1f7caaa1cd2cdfa
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index f4b2095..d19180d 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -577,6 +577,9 @@
 		} else if (err == DMA_UNALIGNED_DEST_ADDR) {
 			arcboot_rom_status_update_bringup(img_sts_base,
 					PROGRESS_BRINGUP_UNALIGNED_DEST_ADDR);
+		} else if (err == DMA_INVALID_IMG_SIZE) {
+			arcboot_rom_status_update_bringup(img_sts_base,
+					PROGRESS_BRINGUP_INVALID_IMG_SIZE);
 		} else if (err == DMA_UNALIGNED_IMG_SIZE) {
 			arcboot_rom_status_update_bringup(img_sts_base,
 					PROGRESS_BRINGUP_UNALIGNED_IMG_SIZE);
diff --git a/arch/gaudi3/arc/dev_dma.h b/arch/gaudi3/arc/dev_dma.h
index 147c7e6..fed69ed 100644
--- a/arch/gaudi3/arc/dev_dma.h
+++ b/arch/gaudi3/arc/dev_dma.h
@@ -10,4 +10,5 @@
 #define MAX_DMA_SIZE_CRC	(MAX_DMA_SIZE + CRC32_SIZE)
 #define OFFSET_64KB		0x10000
 #define ADDR_SIZE_ALIGNMENT	16	/* 16 bytes */
-
+#define MIN_DMA_SIZE		16	/* 16 bytes */
+#define MIN_DMA_SIZE_CRC	(MIN_DMA_SIZE + CRC32_SIZE)
diff --git a/arch/gaudi3/arc/dma.c b/arch/gaudi3/arc/dma.c
index 2f8844d..74a5f73 100644
--- a/arch/gaudi3/arc/dma.c
+++ b/arch/gaudi3/arc/dma.c
@@ -26,6 +26,9 @@
 	if (image_size % MAX_DMA_SIZE_CRC)
 		dma_chunks++;
 
+	if (image_size < MIN_DMA_SIZE_CRC)
+		return DMA_INVALID_IMG_SIZE;
+
 	if ((image_size - (CRC32_SIZE * dma_chunks)) % ADDR_SIZE_ALIGNMENT)
 		return DMA_UNALIGNED_IMG_SIZE;
 
diff --git a/arch/gaudi3/arc/progress_bits.h b/arch/gaudi3/arc/progress_bits.h
index b486457..c8492b2 100644
--- a/arch/gaudi3/arc/progress_bits.h
+++ b/arch/gaudi3/arc/progress_bits.h
@@ -10,127 +10,134 @@
 	PROGRESS_INIT_SPI_BOOT_DESC_DMA_START = 3,
 	PROGRESS_INIT_SPI_BOOT_DESC_CRC_FAIL = 4,
 	PROGRESS_INIT_SPI_BOOT_DESC_UNALIGNED_DEST = 5,
-	PROGRESS_INIT_SPI_BOOT_DESC_UNALIGNED_IMG_SIZE = 6,
-	PROGRESS_INIT_SPI_BOOT_DESC_DMA_ERROR = 7,
-	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 8,
-	PROGRESS_INIT_SPI_BOOT_DESC_HDR_MAGIC_FAIL = 9,
-	PROGRESS_INIT_SPI_BOOT_DESC_VERIFICATION_FAIL = 10,
-	PROGRESS_INIT_SPI_BOOT_DESC_INFO_SET_FAIL = 11,
-	PROGRESS_INIT_SPI_BOOT_DESC_DATA_SET_FAIL = 12,
-	PROGRESS_INIT_SPI_BOOT_DESC_INFO_GET_FAIL = 13,
-	PROGRESS_INIT_SPI_BOOT_DESC_DATA_GET_FAIL = 14,
-	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 15,
-	PROGRESS_INIT_EFUSE_SPI = 16,
-	PROGRESS_REVOKE_LOC = 17,
-	PROGRESS_REVOKE_LOC_0 = 18,
-	PROGRESS_REVOKE_LOC_1 = 19,
-	PROGRESS_KEY_LOC = 20,
-	PROGRESS_KEY_LOC_0 = 21,
-	PROGRESS_KEY_LOC_1 = 22,
-	PROGRESS_VER_LOC = 23,
-	PROGRESS_VER_LOC_0 = 24,
-	PROGRESS_VER_LOC_1 = 25,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_START = 26,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_VERSION_FAIL = 27,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_REVOKED = 28,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 29,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_AUTH_FAIL = 30,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_SUCCESS = 31,
-	PROGRESS_BRINGUP_PCI_PRI = 32,
-	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 33,
-	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 34,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 35,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 36,
-	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 37,
-	PROGRESS_BRINGUP_PCI_PRI_DMA = 38,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 39,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 40,
-	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 41,
-	PROGRESS_BRINGUP_PCI_SEC = 42,
-	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 43,
-	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 44,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 45,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 46,
-	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 47,
-	PROGRESS_BRINGUP_PCI_SEC_DMA = 48,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 49,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 50,
-	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 51,
-	PROGRESS_ARCBOOT_IN_RECOVERY = 52,
-	PROGRESS_ARCBOOT_ROM_TPM = 53,
-	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 54,
-	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 55,
-	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 56,
-	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 57,
-	PROGRESS_BRINGUP_TPM_SPI = 58,
-	PROGRESS_BRINGUP_TPM_SPI_PRI = 59,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 60,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 61,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 62,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 63,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 64,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 65,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 66,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 67,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 68,
-	PROGRESS_BRINGUP_TPM_SPI_SEC = 69,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 70,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 71,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 72,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 73,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 74,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 75,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 76,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 77,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 78,
-	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 79,
-	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 80,
-	PROGRESS_PPBOOT_START = 81,
-	PROGRESS_PPBOOT_FINISH = 82,
-	PROGRESS_BRINGUP_RAM_PRI = 83,
-	PROGRESS_BRINGUP_RAM_PRI_DMA = 84,
-	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 85,
-	PROGRESS_BRINGUP_RAM_SEC = 86,
-	PROGRESS_BRINGUP_RAM_SEC_DMA = 87,
-	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 88,
-	PROGRESS_ARCBOOT_RAM_START = 89,
-	PROGRESS_ARCBOOT_RECOVERY = 90,
-	PROGRESS_ARCBOOT_RECOVERY_DMA = 91,
-	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 92,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 93,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 94,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 95,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 96,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 97,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 98,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 99,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 100,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 101,
-	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 102,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 103,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 104,
-	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 105,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 106,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 107,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 108,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 109,
-	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 110,
-	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 111,
-	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 112,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 113,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 114,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 115,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 116,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 117,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 118,
-	PROGRESS_ARCBOOT_EROM_IMG0 = 119,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 120,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 121,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 122,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 123,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 124,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 125,
-	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 126,
-	PROGRESS_ARCBOOT_EROM_IMG0_START = 127,
-	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 128,
+	PROGRESS_INIT_SPI_BOOT_DESC_INVALID_IMG_SIZE = 6,
+	PROGRESS_INIT_SPI_BOOT_DESC_UNALIGNED_IMG_SIZE = 7,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_ERROR = 8,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 9,
+	PROGRESS_INIT_SPI_BOOT_DESC_HDR_MAGIC_FAIL = 10,
+	PROGRESS_INIT_SPI_BOOT_DESC_VERIFICATION_FAIL = 11,
+	PROGRESS_INIT_SPI_BOOT_DESC_INFO_SET_FAIL = 12,
+	PROGRESS_INIT_SPI_BOOT_DESC_DATA_SET_FAIL = 13,
+	PROGRESS_INIT_SPI_BOOT_DESC_INFO_GET_FAIL = 14,
+	PROGRESS_INIT_SPI_BOOT_DESC_DATA_GET_FAIL = 15,
+	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 16,
+	PROGRESS_INIT_EFUSE_SPI = 17,
+	PROGRESS_REVOKE_LOC = 18,
+	PROGRESS_REVOKE_LOC_0 = 19,
+	PROGRESS_REVOKE_LOC_1 = 20,
+	PROGRESS_KEY_LOC = 21,
+	PROGRESS_KEY_LOC_0 = 22,
+	PROGRESS_KEY_LOC_1 = 23,
+	PROGRESS_VER_LOC = 24,
+	PROGRESS_VER_LOC_0 = 25,
+	PROGRESS_VER_LOC_1 = 26,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_START = 27,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_VERSION_FAIL = 28,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_REVOKED = 29,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 30,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_AUTH_FAIL = 31,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_SUCCESS = 32,
+	PROGRESS_BRINGUP_PCI_PRI = 33,
+	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 34,
+	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 35,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 36,
+	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 37,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 38,
+	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 39,
+	PROGRESS_BRINGUP_PCI_PRI_DMA = 40,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 41,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 42,
+	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 43,
+	PROGRESS_BRINGUP_PCI_SEC = 44,
+	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 45,
+	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 46,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 47,
+	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 48,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 49,
+	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 50,
+	PROGRESS_BRINGUP_PCI_SEC_DMA = 51,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 52,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 53,
+	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 54,
+	PROGRESS_ARCBOOT_IN_RECOVERY = 55,
+	PROGRESS_ARCBOOT_ROM_TPM = 56,
+	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 57,
+	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 58,
+	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 59,
+	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 60,
+	PROGRESS_BRINGUP_TPM_SPI = 61,
+	PROGRESS_BRINGUP_TPM_SPI_PRI = 62,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 63,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 64,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 65,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_INVALID_IMG_SIZE = 66,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 67,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 68,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 69,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 70,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 71,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 72,
+	PROGRESS_BRINGUP_TPM_SPI_SEC = 73,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 74,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 75,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 76,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_INVALID_IMG_SIZE = 77,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 78,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 79,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 80,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 81,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 82,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 83,
+	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 84,
+	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 85,
+	PROGRESS_PPBOOT_START = 86,
+	PROGRESS_PPBOOT_FINISH = 87,
+	PROGRESS_BRINGUP_RAM_PRI = 88,
+	PROGRESS_BRINGUP_RAM_PRI_DMA = 89,
+	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 90,
+	PROGRESS_BRINGUP_RAM_SEC = 91,
+	PROGRESS_BRINGUP_RAM_SEC_DMA = 92,
+	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 93,
+	PROGRESS_ARCBOOT_RAM_START = 94,
+	PROGRESS_ARCBOOT_RECOVERY = 95,
+	PROGRESS_ARCBOOT_RECOVERY_DMA = 96,
+	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 97,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 98,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 99,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 100,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 101,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 102,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 103,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 104,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 105,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 106,
+	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 107,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 108,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 109,
+	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 110,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 111,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 112,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 113,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 114,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 115,
+	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 116,
+	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 117,
+	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 118,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 119,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 120,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 121,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 122,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 123,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 124,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 125,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 126,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 127,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 128,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 129,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 130,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 131,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 132,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 133,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 134,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 135,
 };
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index bf2ff49..2ddc6d8 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -59,6 +59,9 @@
 		} else if (dma_err == DMA_UNALIGNED_DEST_ADDR) {
 			arcboot_rom_status_update_bringup(img_sts_base,
 					PROGRESS_EROM_DMA_UNALIGNED_DEST_ADDR);
+		} else if (dma_err == DMA_INVALID_IMG_SIZE) {
+			arcboot_rom_status_update_bringup(img_sts_base,
+					PROGRESS_EROM_DMA_INVALID_IMG_SIZE);
 		} else if (dma_err == DMA_UNALIGNED_IMG_SIZE) {
 			arcboot_rom_status_update_bringup(img_sts_base,
 					PROGRESS_EROM_DMA_UNALIGNED_IMG_SIZE);
diff --git a/common/boot_desc.c b/common/boot_desc.c
index a0cb8ee..7159724 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -95,6 +95,8 @@
 			arcboot_rom_status_update_bringup(dma_sts_base, PROGRESS_DMA_CRC_FAILED);
 		else if (err == DMA_UNALIGNED_DEST_ADDR)
 			arcboot_rom_status_update_bringup(dma_sts_base, PROGRESS_DMA_UNALIGNED_DEST);
+		else if (err == DMA_INVALID_IMG_SIZE)
+			arcboot_rom_status_update_bringup(dma_sts_base, PROGRESS_DMA_INVALID_IMG_SIZE);
 		else if (err == DMA_UNALIGNED_IMG_SIZE)
 			arcboot_rom_status_update_bringup(dma_sts_base, PROGRESS_DMA_UNALIGNED_IMG_SIZE);
 		else
diff --git a/common/dma.h b/common/dma.h
index 6fd1c72..247e8f0 100644
--- a/common/dma.h
+++ b/common/dma.h
@@ -25,6 +25,7 @@
 	DMA_INVALID_DEST_TYPE,
 	DMA_UNALIGNED_IMG_SIZE,
 	DMA_UNALIGNED_DEST_ADDR,
+	DMA_INVALID_IMG_SIZE,
 	DMA_ERROR_MAX
 };
 
diff --git a/include/arcboot.h b/include/arcboot.h
index 797a5e8..927616d 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -98,6 +98,7 @@
 	PROGRESS_BRINGUP_TOO_BIG,
 	PROGRESS_BRINGUP_CRC_FAIL,
 	PROGRESS_BRINGUP_UNALIGNED_DEST_ADDR,
+	PROGRESS_BRINGUP_INVALID_IMG_SIZE,
 	PROGRESS_BRINGUP_UNALIGNED_IMG_SIZE,
 	PROGRESS_BRINGUP_DMA_ERROR,
 	PROGRESS_BRINGUP_DMA,
@@ -110,6 +111,7 @@
 	PROGRESS_DMA_START,
 	PROGRESS_DMA_CRC_FAILED,
 	PROGRESS_DMA_UNALIGNED_DEST,
+	PROGRESS_DMA_INVALID_IMG_SIZE,
 	PROGRESS_DMA_UNALIGNED_IMG_SIZE,
 	PROGRESS_DMA_ERROR,
 	PROGRESS_DMA_FINISH,
@@ -131,6 +133,7 @@
 	PROGRESS_EROM_DMA_START,
 	PROGRESS_EROM_DMA_CRC_FAILED,
 	PROGRESS_EROM_DMA_UNALIGNED_DEST_ADDR,
+	PROGRESS_EROM_DMA_INVALID_IMG_SIZE,
 	PROGRESS_EROM_DMA_UNALIGNED_IMG_SIZE,
 	PROGRESS_EROM_DMA_ERROR,
 	PROGRESS_EROM_DMA_FINISH,
